### ISSUE TITLE
NullPointerException when no gamepad present.

### ISSUE THREAD

**Issue by cjritola** [jtrfp/terminal-recall#121]
From WDLMaster:
https://gist.github.com/cjritola/9714dcab693c62d39839

**Comment by cjritola** [comment]
Committed possible fixes:
3e61823ed4e56f2fe9143bf7d6f8e0cf7291b67e
2568964713e2425aff2eaeb120adf2f0f22fd113
3033cf1e7304cfa0e2bb76e5fff25eecf007acf6
...does issue persist?

**Comment by WDLmaster** [comment]
The crash's gone but now it does nothing when clicking "New Game". Not even a console entry is written. Hmm… I haven't updated Java in ages. The last update is from around mid-2013. Could this be an issue? I'll update it anyways. If that doesn't help maybe my messing around with custom PODs broke something.

**Comment by WDLmaster** [comment]
[later that day...]

No change. Updated Java and removed the custom POD. 

http://wdlmaster.de/TRCL/log17.html

If I understand that NPE correctly, there should be a new controller / gamepad tab in the config dialog, right? But there is none.

**Comment by cjritola** [comment]
Reproduced the problem :D   
Caused by no controller present. My error; fix in works. Changing bug name to reflect.

**Comment by WDLmaster** [comment]
But there IS a controller Connected to the PC which is also used quite often. It's a Logitech Dual Action. It works in every other application.

**Comment by cjritola** [comment]
That is an important clue; it's like it is expecting XBox 360 Pad instead of simply dealing with what is given. Results in at least 10hrs from now (very late sleep)

**Comment by cjritola** [comment]
30119a9e1708ea10b166fa9e2bc282c927540e72
d049468e12fd912902125985b04937e5c045c204
5056c394eb74ea84c425ba0bf538533d85a555be
0c25386ab81d9e20738b83b4f2ea426daa28ef56
e57971b2d1ebe8cadadcf646df6034101062b16b
fd9be0a02c3ffa2a6bde812e56a4072ad8810380
6798ec6d18021bd2e55e3b4b911b4629b8f29f27

Below is expected output. Keyboard is mandatory, XBox 360 tab is optional or may be one or more other controllers, sticks, and/or head-trackers.
![trcl-controls](https://cloud.githubusercontent.com/assets/1056698/11436911/138416f6-94b7-11e5-91b4-8b6916adf0ef.png)

I'm very interested in making Logitech Dual Action work. Looking further, it appears the LDA appears as a 'stick' instead of a gamepad, so I loosened the input filter to everything other than Keyboard since Keyboard is already supported through the more-mainstream AWT.
http://forums.logitech.com/t5/Logitech-G-Controllers/Dual-Action-Controller-Java-JInput-capabilities/td-p/698492

It'd be great if you could confirm functioning of your LDA. Also I'd appreciate a screenshot of the tab for your LDA after you configure it which I can look at so I may create a default configuration to save users from having to set it up. A TRCL config file might work as an alternative.

**Comment by WDLmaster** [comment]
OK, the Dual Action works! I don't know why it's reported as stick. Maybe a driver issue (?) because in Windows it's simply recognized as gamepad. I'm not sure if a default controller configuration makes any sense because neither the button layout nor the handling of the analog sticks is consistent across gamepads. Furthermore, other people are not necessarily comfortable with my button assignments. Sounds strange but I actually NEVER played TV with a controller. It feels too laggy for my taste. I prefer playing it with keyboard (yes, you heard right). Anyway, now to the config:

![config screen](http://wdlmaster.de/TRCL/config.png)

Z-rotation and Z-axis are the right analog stick.
Y-axis and X-axis are the left (main) analog stick.

I had to invert (set scalar to -1) the analog sticks. For some reason it does not scale the X-axis with my specified factor. But this is needed to pan left/right, which is inverted when no negative factor is used. The other 2 inversions work just fine.

All other buttons work just fine. 

The D-PAD, which is internally handled like a 45-degree-step rotation could be used as weapon select switch but you haven't implemented a weapon select assignment yet.

Button 0 and 1 are most certainly swapped in other controllers. The same applies to 2 and 3. The Dual Action has a counter clockwise layout. Some have a row-based layout, some are clockwise etc. Furthermore, the Dual Action has no analog shoulder buttons while other controllers might have them (which would be another axis).

For some reason, the config dialog shows tabs of seemingly virtual devices that are not physically connected to the PC. It says "multimedia keyboard" but my keyboard is the most basic you can get without ANY additional knobs and buttons for volume and stuff so I have no idea where those entries are coming from. To me it almost seems as all the buttons on my scanner and printers are listed as if they're regular HIDs. Strange.

**Comment by cjritola** [comment]
Are you able to map your scanner's quick-scan button to Fire? (:

Thanks for the screenshot; even if your setup is out of the ordinary it is better than nothing. The controllers have unique input names so I can't reliably guess them.

**Comment by WDLmaster** [comment]
Steam collects user's hardware details but unfortunately, game controllers are not in there. This would've come in handy in this situation so we could at least create a button preset for the 5 (or so) most used controllers. I think the 360 controller is by far the most used. I just compared the button numbers of the 360 to the Dual Action and it's completely different. The Dual Action goes counter clockwise starting at the left button, the 360 starts on the lower button and goes in a Z pattern. Furthermore, the D-PAD on 360 has 2 axes while the Dual Action has just one axis. I think I even saw gamepads where the D-PAD has 4 separate buttons! Aw man… 

in short: a standard button config is useless. Maybe you can implement a quick-assign dialog where you just have the available options of TRCL (fire, afterburner …) and the user simply presses his/her preferred button to advance to the next entry. I'm not sure the simple, non-programmer people out there know what to do with "scalar" and "offset" values. 

Oh, btw, I suggest you implement a (possibly user definable) dead-zone for all analog axes (around 10% I'd say). The analog sticks never really rest exactly at zero. Especially not if you quickly let go the stick.
